### DISCUSSION TITLE
Proposed New Update Tool and Update Method

### DISCUSSION THREAD

**Discussion by mcguirepr89** [mcguirepr89/BirdNET-Pi#282] [Q&A]
Updating sucks! :)

I really want to make it better, but have found that figuring out what will work and always work has been a bit of a moving target. As soon as I think I have a method figured out, I learn that it does not work for everyone.

I wanted to ask this community from those of you with professional application development and other collaboration/version-control type stuff how I might be able to implement better, more reliable updates.

My best regards as always,
Patrick

**Comment by ehpersonal38** [comment]
It doesn't even make sense to me why the current update code doesn't work for some people. I just tested it a bunch and it worked :(

**Reply by airigami** [reply]
Well, I don't know how many other people have run into upgrade issues. in my case, I looked for instructions that told me how to update and found outdated info. If I had just found the "update" button in the web interface, I may never have looked further, avoiding trouble completely. (Or maybe I still would have had a problem if it is a bad card that took down my installation.) I don't know that the upgrade mechanism can be blamed. My problem can be avoided in the future with a change to the documentation (which I believe already happened). And on that note ...

I can't be very helpful on the development side of things. I haven't coded anything in years. But maybe I can assist on documentation if that's needed. I don't know what is needed. My random difficulty aside, the system just kinda works. I don't know how much documentation anyone wants.

**Comment by jmherbst** [comment]
My 2cents 🪙 

### Git method (current)
In order to continue the `git stash; git pull; git stash pop` we need to refactor all configurable attributes into proper non-git-tracked configuration files (aka birdnet.conf)

Some problems I've noticed are that things are configurable, but only directly in the git-tracked scripts - Like mentioned in this comment https://github.com/mcguirepr89/BirdNET-Pi/pull/268#issuecomment-1123843497

Everything configurable needs to be pulled into their own ENV vars, and sourced properly - Once that is done, then you can remove the `git stash` portions of the upgrade.

This should remove a lot of the merge-conflicts encountered in this method.  Any merge-conflicts that do arise, can be chalked up to "You shouldn't modify those files locally. please use birdnet.conf variables" 

### Packaged + Versioned releases
We'll need to figure out a way to package the application up in a way that is repeatable and install'able on desired OS's. This will also require a lot of the "configuration refactor" from the *Git Method* above...

With this you can add automations around the package building and versioned releasing.  With versioned releases you'll also gain a sort of "contract" with users where small changes can be updated easily and not require any major changes to their systems, but larger/Major changes are your way of saying "This is a major overhaul of the application or its dependencies and these steps are needed to be taken prior/after upgrading"

**Reply by mcguirepr89** [reply]
>Everything configurable needs to be pulled into their own ENV vars, and sourced properly - Once that is done, then you can remove the git stash portions of the upgrade.

I agree and have been lazy. I'll do my best to see to this~ the "Advanced Settings" page is about to be a big one!

**Comment by phrozen77** [comment]
Being a new "BirdNET'er" - first of all: thank you for what seems to be an enormous effort!

Secondly, i surely can be of assistance in reporting and debugging any kind of issues when updating - should any arise...

Is there already any kind of documentation re: updating an installation?

IMHO - no offense meant - standard procedures like installing / updating / backups should be documented (if even roughly, someone else can fill in prose as seen fit) but also all configuration options should be there.

I'd gladly help filling the docs - as i said, even if it's just a few rough skeletons / "red links" in the Wiki...

**Reply by mcguirepr89** [reply]
Thanks for your kind words and offerings.
>Secondly, i surely can be of assistance in reporting and debugging any kind of issues when updating - should any arise...

:+1: 
>Is there already any kind of documentation re: updating an installation?

Only the [small excerpt on the ReadMe](https://github.com/mcguirepr89/BirdNET-Pi#updating)

>IMHO - no offense meant - standard procedures like installing / updating / backups should be documented (if even roughly, someone else can fill in prose as seen fit) but also all configuration options should be there.
>
>I'd gladly help filling the docs - as i said, even if it's just a few rough skeletons / "red links" in the Wiki...

Please feel free to start creating some documentation for the project as you get familiar with how it functions! I know of many articles I wish were available to users which I have neglected to write while things have been in development -- (there has been a LOT of documentation come and go, even an entire other now-redirected domain, https://birdnetwiki.pmcgui.xyz when the system was _BirdNET-system_, named after `systemd` since it was originally just an x86 installer, then became aarch64 after a while -- just some history for you :wink:) -- but I don't always go back and make the documentation after development around a thing has settled. Also, there are a LOT more users now than there were when I started, so development is so much quicker now! It's _pretty, pretty, pretty_ cool, as Larry David might put it.

**Comment by mcguirepr89** [comment]
I've thought a bit about this, and had this idea (could be half-baked, could be usable, hard to say at this point):

If there were two Update options, the user could decide whether they want to keep local changes (`stash`/`pull`/`stash pop`) or to keep their system up to date with `main` (`checkout -f`/`pull -f`/`update_birdnet.sh`).

The options could be along the lines of "Update (Keep local changes)" "Upgrade (discard local changes and get up to date with `main`)"

Ok, half-baked is my verdict after reading it over, but worth posting since it could spark another idea elsewhere.

**Reply by airigami** [reply]
What sorts of local changes are people making outside of configuration changes? I think it's only fair that if someone wants a smooth update, they stick with a supported package? If a change you made is so awesome that you don't want to lose it, chances are others can use it also and it should be folded back into the distribution. If everyone keeps up to date, upgrades should be really easy (comparatively). 

Of course, I do realize keeping up to date isn't always as easy if you're using it out in the field and not always connected to the internet.

**Reply by ehpersonal38** [reply]
I actually like this solution, although maybe there'd be just one Update button, but after you click it it would check if you have local changes and if so, then prompt you with the option to clear them or keep them.

**Reply by mcguirepr89** [reply]
@ehpersonal38
I always appreciate your UI input and think that is a :sunglasses: + :microscope: approach --

Thoughts on the efficacy of `update_birdnet_snippets.sh` after a bit of use?

@airigami, as @jmherbst mentioned, https://github.com/mcguirepr89/BirdNET-Pi/pull/268#issuecomment-1123843497 are among the customizations in use -- added to the list can be aesthetic and web page language changes. But you are not wrong that it is not reasonable to _support_ customizations -- I do, however, want to _allow_ customizations since I find user-access to the code integral to what makes BirdNET-Pi such a powerful tool, while also being able to obtain new features and security patches.

**Reply by Svardsten53** [reply]
I have made translations into Swedish of the pages and some other adjustments of a couple of pages and have also added my own link page.

I have written some scripts that do this and in order for the updates to work properly, I restore the original files in the directories homepage and scripts (except birds.db) before I run an Update. After the update, I run the customization scripts again. Then I get a combination of the latest updates and my own translations and customizations. So far it seems to work well.

**Reply by ehpersonal38** [reply]
> Thoughts on the efficacy of `update_birdnet_snippets.sh` after a bit of use?

I think it's a great script! 😄 I can't really think of a better solution.

**Comment by Nachtzuster** [comment]
Another cause of local changes is BirdNET itself: 
- injecting shebangs on the first line of the *.py scripts at install time
- overwriting labels.txt from the settings page when changing the language.

In case of the shebangs, I believe changing just the first line of a python file in this repo, will cause a conflict when users update their BirdNET installation via the GUI.

In general, it would be a good if a vanilla BirdNET could run from unmodified code.

**Reply by jmherbst** [reply]
I noticed this earlier today as well. I'll try and get something up to correct this.
